### ISSUE TITLE
[1.19.3] Custom sky rotation is broken

### ISSUE THREAD

**Issue by SeaOfPixels** [sp614x/optifine#7235]
## Description of Issue
At certains times in the day, the custom sky rotation accelerates to hyperspeed. This doesn't happen with the same sky properties in 1.19.2. The speed eventually returns to normal.

## OptiFine Version
1.19.3 I2 pre2

https://user-images.githubusercontent.com/62183036/213954787-39a3a7f4-5d4c-4c53-8c1d-300be91f1c38.mp4

**Comment by SeaOfPixels** [comment]
In the latest release of my pack I set the speed to 1.0 as a temporary fix for this bug, but the sky still speeds up towards the end of the night and I don't understand why. It doesn't do this in 1.12 even though I thought they used the same rotation method. Rly frustrating to not be able to fix this on my end.

**Comment by UsernameGeri** [comment]
You have the axis specified incorrectly. Write `axis=0.0 -1.0 0.0` and it will solve the issue.

**Comment by SeaOfPixels** [comment]
This fixes the speeding up and slowing down, but it doesn't fix the sky still using the old literal rotation method rather than simulated.

**Comment by UsernameGeri** [comment]
Old literal rotation method? Simulated? What?

**Comment by SeaOfPixels** [comment]
I explained it in the post above, the way the sky rotates changed between OptiFine 1.19.2 and OptiFine 1.19.3, even with your axis fix. The rotation method is now back to what was used all the way back in 1.12 now.

**Comment by UsernameGeri** [comment]
I tested 1.12.2, 1.19.2-3- and 4 and it all works the exact same way. Can you show an example where the sky behaves differently between these version? Also I still have no idea what you mean by "simulated" and "back and forth". I think you're a bit confused about this, I don't think the sky rotation has ever been changed since its integration from mcpatcher to optifine.

**Comment by SeaOfPixels** [comment]
Here is a comparison:

1.19.2, sky simulates a rotation by swiveling back and forth:
https://github.com/sp614x/optifine/assets/62183036/0deef3b0-0c5b-44cf-93fa-e05084798615

1.19.4, sky literally rotates in a circle:
https://github.com/sp614x/optifine/assets/62183036/37d1b4fc-e8eb-415e-8add-ef61ef99e435

**Comment by UsernameGeri** [comment]
How does this swivelling look in real time? I've never seen such behavior before. Can you also share the properties file for this sky?

**Comment by SeaOfPixels** [comment]
They're the properties in the original post. Also this is the best I can show it within the constraints of video file size, I assure you it swivels like this without adding time.

**Comment by roro1506HD** [comment]
After looking more closely to the source code of 1.19.2- OptiFine and 1.19.3+, there is nothing that changed regarding all that.
What I looked and tested though, is that 1.19.3 was when JOML has been introduced in the code, and this is what is causing all this. 

I tested by bringing back some old Mojang's math library classes, tweaked them a bit to fix issues and I can affirm that the JOML calculation is what is causing all that mess. 
The old math library and the new are definitely not doing the same calculations.

I took out the necessary math to compute the rotation and put it all in a method, to save some time to sp614x if he wants to use it: https://pastebin.com/pSWqZVh7
This method can be used in replacement of `PoseStack#rotateDeg`:
Current: 
```java
matrixStackIn.rotateDeg(360.0F * (angleDayStart + celestialAngle * this.speed), this.axis[0], this.axis[1], this.axis[2]);
```
Replacement:
```java
this.replicateOldMulPoseBehavior(
    matrixStackIn,
    (float) Math.toRadians(360.0F * (angleDayStart + celestialAngle * this.speed)), // The main difference is this being radians instead of degrees
    this.axis[0], 
    this.axis[1], 
    this.axis[2]
);
```
---
### TL;DR
A math thing has changed in 1.19.3 that is causing all this, despite OptiFine not changing anything

**Comment by SeaOfPixels** [comment]
Hoping OptiFine is able to account for and fix this on their end, thank you for looking into this.

**Comment by BlockyTheDev** [comment]
I hope this gets fixed soon.

**Comment by Jiingy** [comment]
@sp614x Considering how prominent it is for resource pack users, it may be good to fix soon. A bug tester has provided an example of how to fix the issue here.

> After looking more closely to the source code of <1.19.2 OptiFine and 1.19.3+, there is nothing that changed regarding all that. What I looked and tested though, is that 1.19.3 was when JOML has been introduced in the code, and this is what is causing all this.
> 
> I tested by bringing back some old Mojang's math library classes, tweaked them a bit to fix issues and I can affirm that the JOML calculation is what is causing all that mess. The old math library and the new are definitely not doing the same calculations.
> 
> I took out the necessary math to compute the rotation and put it all in a method, to save some time to sp614x if he wants to use it: https://pastebin.com/pSWqZVh7 This method can be used in replacement of `PoseStack#rotateDeg`: Current:
> 
> ```java
> matrixStackIn.rotateDeg(360.0F * (angleDayStart + celestialAngle * this.speed), this.axis[0], this.axis[1], this.axis[2]);
> ```
> 
> Replacement:
> 
> ```java
> this.replicateOldMulPoseBehavior(
>     matrixStackIn,
>     (float) Math.toRadians(360.0F * (angleDayStart + celestialAngle * this.speed)), // The main difference is this being radians instead of degrees
>     this.axis[0], 
>     this.axis[1], 
>     this.axis[2]
> );
> ```
> 
> ### TL;DR
> A math thing has changed in 1.19.3 that is causing all this, despite OptiFine not changing anything

**Comment by UsernameGeri** [comment]
I wonder how reliable this fix would be. People will continue to use incorrect axes, would this fix account for that?

**Comment by Jiingy** [comment]
Pinning this issue as a solution has been provided.

**Comment by Bhrol** [comment]
Is the problem still active ? I'm experiencing the same type of bug right now

**Comment by UsernameGeri** [comment]
From what I can tell, this happens because the skybox is too close to the camera/player. In the Celestial mod, you can create your own custom vertices for skyboxes, and if you create vertices that are too close, this same bobbing will occur. If the vertex is very far away, the bobbing is un-noticeable.

**Comment by sp614x** [comment]
The sky rotation is not working correctly when the axis is not normalized (0, -0.2, 0 instead of 0, -1, 0).
In 1.19.2 the sky rotates back and forth with different speeds depending on the time of day. This is a bug and not intended, the sky should rotate uniformly around the axis.
In 1.19.4 and later the rotation is correct, but the bobbing effect is distorted.
To fix both problems the axis should be normalized (x² + y² + z² = 1).
Custom Sky should be fixed to automatically normalize the axis and avoid the side effects.

**Comment by sp614x** [comment]
If the old behavior is still needed, Custom Sky may add a compatibility flag "legacyRotation" or similar that doesn't normalize the axis and uses the old math to do the rotation.
A better solution for a back and forth effect would be to add some cos/sin options so the movement can be randomized a bit, for example:
```rotation=linear|legacy|sin|cos```

**Comment by sp614x** [comment]
Looks like normalization is not required: https://github.com/sp614x/optifine/blob/master/OptiFineDoc/doc/sky.properties#L107
But it is not defined how the non-normalized axis rotation should behave.

**Comment by lowercasebtw** [comment]
> Looks like normalization is not required: https://github.com/sp614x/optifine/blob/master/OptiFineDoc/doc/sky.properties#L107
> But it is not defined how the non-normalized axis rotation should behave.

Check discord #general-vip

**Comment by sp614x** [comment]
As I see it when the axis is normalized all is good and all versions work identically.
With non-normalized axis the sky movement in 1.17-1.19.2 is non-linear (back and forth), which some RP use as creative effect, but it generally doesn't look broken (no sky bobbing).
In 1.19.4+ versions the non-normalized axis leads to visually broken sky rotation (bobbing, speed). 

Probably the best solution is to update the documentation with description of the effects.
Then fix the handling of non-normalized axis to behave like 1.17-1.19.2.

Then RPs that want smooth rotation can use normalized axis and RPs that want to use the creative back and forth can use non-normalized axis.

**Comment by UsernameGeri** [comment]
I'm pretty sure nobody ever used non-normalized axis as a "creative effect", becuase it's literally the same value used in every pack. At this point I've seen *hundreds* or sky packs, and without fault, every single one had the same exact broken value.
This, to be specific: axis=0.0 -0.2 0.0
I think this is just a simple case of people copying each others' packs over and over. Someone early on made this broken config, probably because they didn't understand how the axis worked, and it's been spreading around ever since.

Edit.: When I say `-0.2` is "broken", I mean that in the current working conditions of optifine it's broken, because it's not being normalized.
